server/audit/views.py

```diff
@@ -423,20 +426,124 @@ def get_permissions(self):
         return [permission() for permission in permission_classes]
 
     def list(self, request):
-        org_id = request.GET.get("organization", None)
+        org_id = request.GET.get('organization', None)
         bins_to_recommend = list(BinToSK.objects.filter(init_audit__organization=org_id).values('Bin').
                                  annotate(total=Count('Bin')).values('Bin', 'total').order_by('-total')[:5])
 
         parts_to_recommend = list(
             Record.objects.filter(bin_to_sk__init_audit__organization=org_id).values('Part_Number').annotate(
                 total=Count('Part_Number'))
-            .values('Part_Number', 'total').order_by('total').order_by('-total')[:5])
+                .values('Part_Number', 'total').order_by('total').order_by('-total')[:5])
 
         items_to_recommend = list(
             Record.objects.filter(bin_to_sk__init_audit__organization=org_id, flagged=True).values('item_id').annotate(
-                total=Count('item_id'))
-            .values('item_id', 'total').order_by('-total')[:5])
+                total=Count('item_id')).values('item_id', 'total').order_by('-total')[:5])
+
+        # Recommend an Ad-hoc audit : random selection od bins or parts
+        all_bins = get_values(list(BinToSK.objects.filter(init_audit__organization=org_id).values('Bin').distinct()),
+                              'Bin')
+        all_parts = get_values(
+            list(Record.objects.filter(bin_to_sk__init_audit__organization=org_id).values('Part_Number').distinct()),
+            'Part_Number')
+
+        random_items_to_recommend = get_random_items(all_bins, all_parts, org_id, 2)
+
+        # Recommend items with high criticality
+        high_criticality_items = list(
+            Item.objects.filter(organization=org_id, Criticality='High').values(
+                'Item_Id'))
 
         data = {'bins_recommendation': bins_to_recommend, 'parts_recommendation': parts_to_recommend,
-                'items_recommendation': items_to_recommend}
+                'items_recommendation': items_to_recommend, 'random_items': random_items_to_recommend,
+                'item_based_on_category': high_criticality_items}
         return Response(data)
+
+
+def get_values(list_of_dict, key):
+    return [val[key] for val in list_of_dict]
+
+
+def get_random_items(all_bins, all_parts, org_id, frequency_of_bins_or_parts):
+    random_bins_or_parts = random.sample(["bins", "parts"], 1)[0]
+    if random_bins_or_parts == "bins":
+        print('bins')
+        random_bin = random.sample(all_bins, frequency_of_bins_or_parts)  # increase if we want more items
+        return list(
+            Item.objects.filter(organization=org_id, Bin__in=random_bin).values(
+                'Item_Id'))
+    else:
+        print('part')
```

Comment:
Is this required? 
If not, please remove it

Reply:
fixed